Refactor binary builder

TisVictress opened the pull request:
Remove binary builder code from compilation

sophiewigmore commented:
@TisVictress whats the reason for this PR being a draft still?

sophiewigmore commented on dependency/actions/compile/entrypoint/extensions/extensions-8.1.json:
```diff
@@ -0,0 +1,242 @@
+[
```
why don't we need extensions for 8.0 anymore?

sophiewigmore commented on dependency/actions/compile/bionic.Dockerfile:
```diff
@@ -1,73 +1,65 @@
 FROM ubuntu:18.04
```
Per https://github.com/paketo-buildpacks/php-dist/issues/486, can we compile on our stack image? Same for Jammy

sophiewigmore commented on dependency/actions/compile/entrypoint/main.go:
```diff
@@ -0,0 +1,480 @@
+
+  if outputDir == "" {
+	  fail(errors.New(`missing required input "output-dir"`))
+  }
```
The flag we run with should be `--outputDir`

sophiewigmore commented on dependency/actions/compile/entrypoint/main.go:
```diff
@@ -0,0 +1,480 @@
+  )
+
+  flag.StringVar(&version, "version", "", "PHP dependency version")
+  flag.StringVar(&outputDir, "output-dir", "", "Output directory for PHP dependency artifact")
```
The flag we run with should be `--outputDir`

sophiewigmore commented on dependency/actions/compile/entrypoint/main.go:
```diff
@@ -0,0 +1,480 @@
+  }
+}
+
+func copyFiles(dependencyDir string) error {
```
To the functions, can you add a comment for each just giving a description of what its for?

TisVictress replied:
@sophiewigmore I'll try to give this function a more descriptive name. I think adding a ton of comments to explain what a function does is sort of a tell that the function name is not useful. Would that be okay?

sophiewigmore replied:
this request isn't specific to this function. In general the compilation code is pretty complicated and if ever have to change things, it would help be helpful to have a bit of context on why things are the way they are

sophiewigmore commented on dependency/actions/compile/entrypoint/main.go:
```diff
@@ -0,0 +1,480 @@
+	return nil
+}
+
+func installLua(extensionData ExtensionsData) error {
+	err := downloadAndExtract(
+		extensionData.Name,
+		fmt.Sprintf("http://www.lua.org/ftp/lua-%s.tar.gz", extensionData.Version),
+	)
+	if err != nil {
+		return err
+	}
+
+	dir := fmt.Sprintf("%s-%s", extensionData.Name, extensionData.Version)
+	commands := []CommandWithArgs{
+		CommandWithArgs{"make", []string{"linux", "MYCFLAGS=-fPIC"}},
+		CommandWithArgs{"make", []string{"install"}},
+	}
+
+	for _, command := range commands {
+		err = runCommand(dir, command.Command, command.Args...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```
are all of the `installX` functions the same? Can we abstract that into a separate function to cut down on duplication?

TisVictress replied:
@sophiewigmore Yes and no... some require calling `cmake` or `configure` but the options passed to those commands are different or some of the commands require using different flags.

TisVictress commented:
@sophiewigmore I didn't get a chance to add some comments but I split the main file out into `internal` to make it easier to read. The code still needs to be refactored a bit more to make it more simpler (ie running commands). It now uses the paketo full builders.
The tarballs are not identical to the current output but they are similar but I didn't get a chance to test the latest changes in the buildpack.

The `InstallerFactory` has methods that match the names of the PHP extensions so the utils can use golang's `relect` pkg to create installers for each extension. I split the extensions out into separates files as well, one for native modules and the other for Pecl extensions. 

Compilation works for jammy and bionic.

sophiewigmore commented:
> The code still needs to be refactored a bit more to make it more simpler (ie running commands).

Are you specifically looking for feedback on that? Or is it in progress?


> The tarballs are not identical to the current output but they are similar but I didn't get a chance to test the latest changes in the buildpack.

This is pretty important so we can verify that we haven't changed any behaviour with the new dependencies.

sophiewigmore commented on dependency/actions/compile/entrypoint/extensions/extensions-8.1.json:
```diff
@@ -0,0 +1,248 @@
+      "klass": "PHPIRedisRecipe"
+    },
+    {
+      "name": "oci8",
+      "version": "3.0.1",
+      "md5": "641ebceb483fa2d95e79aea99b1350de",
+      "klass": "OraclePeclRecipe"
+    },
+    {
+      "name": "pdo_oci",
+      "version": "nil",
+      "md5": "nil",
+      "klass": "OraclePdoRecipe"
+    },
+    {
+      "name": "amqp",
+      "version": "1.11.0",
+      "md5": "3a0638223e320fc21f3b079198c24f12",
+      "klass": "AmqpPeclRecipe"
+    },
+    {
+      "name": "oci8",
+      "version": "3.2.1",
+      "md5": "309190ef3ede2779a617c9375d32ea7a",
+      "klass": "OraclePeclRecipe"
```
`oci8` is duplicated here. Is that intentional?

sophiewigmore commented on dependency/actions/compile/entrypoint/extensions/extensions-8.1.json:
```diff
@@ -0,0 +1,248 @@
+      "klass": "AmqpPeclRecipe"
+    },
+    {
+      "name": "oci8",
+      "version": "3.2.1",
+      "md5": "309190ef3ede2779a617c9375d32ea7a",
+      "klass": "OraclePeclRecipe"
```
`oci8` is duplicated in this file. Is that intentional? I think we probably only want one.

sophiewigmore replied:
we're also missing `ioncube` (referencing the original PHP 8.1 extensions file - https://github.com/cloudfoundry/buildpacks-ci/blob/master/tasks/build-binary-new/php81-extensions-patch.yml)

TisVictress replied:
> Is that intentional?

@sophiewigmore I noticed this but I didn't want to remove it because I wasn't sure if this was intentional. Also, the versions of these extensions seem old.

sophiewigmore commented on dependency/actions/compile/bionic.Dockerfile:
```diff
@@ -1,73 +1,75 @@
+USER ${cnb_uid}:${cnb_gid}
 
-RUN bundle install
+RUN apt-get update && \
+  apt-get -y install \
+  autoconf \
+  automake \
+  bison \
+  build-essential \
+  cmake \
+  curl \
+  firebird-dev \
```
If we're compiling on the stack image we shouldn't need to install all of these things because a good number of them should already be there. There may be a few we still need to install

TisVictress replied:
I removed packages that do not need to be reinstalled but some remained.

sophiewigmore commented on dependency/actions/compile/jammy.Dockerfile:
```diff
@@ -1,73 +1,74 @@
-ADD ./binary-builder /binary-builder
-WORKDIR /binary-builder/cflinuxfs4
-RUN bundle install
+RUN apt-get update && \
+  apt-get -y install \
+  autoconf \
+  automake \
+  bison \
+  build-essential \
```
Same comment as on the bionic Dockerfile - If we're compiling on the stack image we shouldn't need to install all of these things because a good number of them should already be there. There may be a few we still need to install

sophiewigmore replied:
@TisVictress this comment still stands.

sophiewigmore replied:
This comment still stands :/ a lot of these packages are on the Full jammy build image.

sophiewigmore commented on dependency/actions/compile/entrypoint/main.go:
```diff
@@ -0,0 +1,311 @@
+
+	var extFile string
+	if strings.Contains(version, "8.0") {
+    extFile = "/tmp/entrypoint/extensions/extensions-8.1.json"
```
I might be missing something but why is this path hardcoded to `/tmp`? Where does that path come from?

TisVictress replied:
@sophiewigmore The Dockerfile places this in /tmp.../
Do you think it makes more sense to just copy to the working dir?

sophiewigmore commented:
> The InstallerFactory has methods that match the names of the PHP extensions so the utils can use golang's relect pkg to create installers for each extension. I split the extensions out into separates files as well, one for native modules and the other for Pecl extensions.

In general the new split has made things clearer, well done.
You have the most context on how the code comes together so I'll defer to you for an answer, but it still feels like there's a lot of duplication. For example, in `internal/extension_installer.go`, the Pecl extensions all look similar for their functions except `Install`.
```
type Redis struct {
	peclExt PeclExtension
}

func (r Redis) DepName() string {
	return r.peclExt.DepName()
}

func (r Redis) DepVersion() string {
	return r.peclExt.DepVersion()
}

func (r Redis) DepSha() string {
	return r.peclExt.DepSha()
}

func (r Redis) URL() string {
	return r.peclExt.URL()
}

func (r Redis) WorkingDir() string {
	return r.peclExt.WorkingDir()
}
```
Would it be possible to achieve some kind of method override pattern, were `PeclExtension` type is anonymous inside of each extension type like:
```
type Redis struct {
	PeclExtension
}
```
and then, we can use the common `PeclExtension` methods like `URL()` and `WorkingDir()`, but override `Install()` and any others that are different? I think we could use that same idea for the various "types" of extensions you have here. What do you think?

sophiewigmore commented:
small nit - I noticed that the `dep.tgz.checksum` file that gets output doesn't include the `algorithm` in the file content

sophiewigmore commented:
Just compiled 8.0.27 (bionic) and 8.1.15 (bionic and jammy) with the code here and put them into the buildpack. The integration tests here fail with:

```
        [builder] failed to evaluate symlink /layers/paketo-buildpacks_php-dist/php/lib/libgpg-error.so: lstat /layers/paketo-buildpacks_php-dis
t/php/lib/lib: no such file or directory
            [builder] ERROR: failed to build: exit status 1
            ERROR: failed to build: executing lifecycle. This may be the result of using an untrusted builder: failed with status code: 51
```
during the step during the build phase where we try to install (and untar) the dependency.

TisVictress commented:
> What do you think?

@sophiewigmore  Each extension type is an `Installer` and needs to implement the `Installer` interface. I don't think this can be done with anonymous types because the InstallerFactory expects each type to have a specific set of methods.

TisVictress commented:
> Are you specifically looking for feedback on that? Or is it in progress?

Both

sophiewigmore commented:
@TisVictress have you had a chance to test out these changes in the buildpack? That's how I discovered the failures last time

sophiewigmore commented on dependency/del.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
What is this script for?

TisVictress replied:
Its a test script that got committed by accident. I've deleted this.

sophiewigmore commented on dependency/actions/compile/bionic.Dockerfile:
```diff
@@ -57,17 +57,15 @@ RUN apt-get update && apt-get install -y \
       ruby-dev \
       rubygems \
       snmp-mibs-downloader \
+      software-properties-common \
```
I can't comment on the specific line, but bionic compilation here is being run on the jammy base image, and the jammy compilation is being run on the stack image (ideally what we want). Is there an intentional reason for this disparity?

TisVictress replied:
Both are now based on paketo stack images

sophiewigmore commented on dependency/actions/compile/entrypoint/extensions/extensions2-8.1.json:
```diff
@@ -0,0 +1,147 @@
+{
```
Why is this named `extensions2`? Do we need this file? Also the extension file for 8.2 is missing

TisVictress commented:
> @TisVictress have you had a chance to test out these changes in the buildpack? That's how I discovered the failures last time

Yes...I used the integration tests and they were green.

robdimsdale reviewed (approved):
In general I think this looks great. Thanks for all the hard work @TisVictress - I appreciate you sticking with it for so long!

My review has essentially only been around writing Golang / using `packit`. I don't have enough context on the php extensions to have an opinion on any of those.

I have a couple of minor suggestions about improved error handling and resiliency (inline). But I don't think any of them should block the merging of this PR.

TisVictress commented on dependency/actions/compile/entrypoint/internal/utils.go:
```diff
@@ -0,0 +1,201 @@
+	extFile := fmt.Sprintf("/tmp/entrypoint/extensions/extensions-%s.json", majorVersion)
+
+	var extensions ExtensionsData
+	jsonBlob, _ := os.ReadFile(extFile)
```
That was a mistake. Fixed :)

sophiewigmore commented on dependency/actions/compile/bionic.Dockerfile:
```diff
@@ -1,73 +1,79 @@
 
-RUN bundle install
+RUN apt-get update && \
+  apt-get -y install \
+  automake \
+  build-essential \
+  bundler \
+  cmake \
+  curl \
```
I am still wondering why we need to install this entire list of packages if were building on the stack

TisVictress replied:
Oh yes sorry @sophiewigmore I'll see what can be removed

sophiewigmore replied:
there may be packages we need here that aren't in the stack image though so might need to do a cross-check against the build image receipt

sophiewigmore commented on dependency/actions/compile/entrypoint/extensions/extensions-8.2.json:
```diff
@@ -0,0 +1,160 @@
+      "md5": "5feec07108cf7898bd0591ee54b572cd"
+    },
+    {
+      "name": "psr",
+      "version": "1.2.0",
+      "md5": "07695da8376002babbce528205decf07"
+    },
+    {
+      "name": "solr",
+      "version": "2.6.0",
+      "md5": "ac5e1e1fbf28e0c543a52c533c967634"
```
Looking at the list of extensions here, some extensions that were in the old extensions-manifests/extensions-8.2.yml file are missing, like `pdo_oci` and `oci8`, and possibly others. I didn't check 8.0 and 8.1 but we need to make sure we're not missing extensions

TisVictress replied:
@sophiewigmore I removed them because it doesn't seem that they are actually getting installed. For example, there is a check for oci8 that seems to always return false for the currently supported versions (https://github.com/cloudfoundry/binary-builder/blob/main/recipe/php_meal.rb#L202). And some extensions were causing `make test` to fail, because the test wasn't expecting to see these extensions pop up. I'll rerun the test and post the failing extensions here. If you think I should add them back, I can update the current tests to include them.

sophiewigmore replied:
@TisVictress that's strange... hmmm. It could be worthwhile just double checking if the versions of PHP in the buildpack.toml now have those extensions or not. If not then we're probably OK to leave them out, but I'm worried to stop supporting them if they do somehow end up available in the versions of PHP we currently ship, you know?

sophiewigmore replied:
Any update on this?

sophiewigmore commented on dependency/test/jammy.Dockerfile:
```diff
@@ -1,9 +1,11 @@
-FROM ubuntu:22.04
+FROM paketobuildpacks/build-jammy-full
```
We're still installing a whole bunch of packages that are on the stack image (starting on line 10). Can we please pare this down?

sophiewigmore commented:
@TisVictress btw I was able to compile 8.1.21 on bionic, but 8.2.8 takes a really long time and errors out on my machine. Are you able to do it on your machine?

TisVictress commented:
> @TisVictress btw I was able to compile 8.1.21 on bionic, but 8.2.8 takes a really long time and errors out on my machine. Are you able to do it on your machine?

@sophiewigmore No I haven't had any issues. Would you mind trying again?
